File: Everything-app-frontend/everything_app/lib/models/contract.dart
/// Eine wiederkehrende Zahlung - Abo, Miete, Versicherung, aber auch das Gehalt.
///
/// [frequency] ist das Etikett für die Gruppierung, [intervalDays] der gemessene
/// Abstand; jede Datumsrechnung im Backend läuft über letzteren.
class Contract {
  final int id;
  final String name;
  final String category;
  final String? subcategory;

  /// 'INCOME' oder 'EXPENSE'.
  final String direction;

  /// Immer positiv - die Richtung steckt in [direction].
  final double amount;

  final String frequency;
  final int? intervalDays;
  final DateTime? lastBookingDate;
  final DateTime? nextDueDate;

  /// "Erkannt aus N Buchungen" - macht nachvollziehbar, woher der Vertrag kommt.
  final int occurrenceCount;

  final bool active;
  final DateTime? cancelledAt;

  /// `false` heißt: vom Nutzer angelegt oder korrigiert. Die Erkennung fasst ihn
  /// dann nicht mehr an.
  final bool detectedAutomatically;

  /// Auf einen Monat normalisiert, damit sich Verträge verschiedener Rhythmen
  /// summieren lassen.
  final double? monthlyAmount;

  Contract({
    required this.id,
    required this.name,
    required this.category,
    this.subcategory,
    required this.direction,
    required this.amount,
    required this.frequency,
    this.intervalDays,
    this.lastBookingDate,
    this.nextDueDate,
    this.occurrenceCount = 0,
    this.active = true,
    this.cancelledAt,
    this.detectedAutomatically = true,
    this.monthlyAmount,
  });

  factory Contract.fromJson(Map<String, dynamic> json) {
    return Contract(
      id: json['id'],
      name: json['name'] ?? '',
      category: json['category'] ?? 'Sonstiges',
      subcategory: json['subcategory'],
      direction: json['direction'] ?? 'EXPENSE',
      amount: (json['amount'] as num?)?.toDouble() ?? 0,
      frequency: json['frequency'] ?? 'MONTHLY',
      intervalDays: (json['intervalDays'] as num?)?.toInt(),
      lastBookingDate: json['lastBookingDate'] != null
          ? DateTime.parse(json['lastBookingDate'])
          : null,
      nextDueDate:
          json['nextDueDate'] != null ? DateTime.parse(json['nextDueDate']) : null,
      occurrenceCount: (json['occurrenceCount'] as num?)?.toInt() ?? 0,
      active: json['active'] ?? true,
      cancelledAt:
          json['cancelledAt'] != null ? DateTime.parse(json['cancelledAt']) : null,
      detectedAutomatically: json['detectedAutomatically'] ?? true,
      monthlyAmount: (json['monthlyAmount'] as num?)?.toDouble(),
    );
  }

  Map<String, dynamic> toJson() => {
        'id': id,
        'name': name,
        'category': category,
        if (subcategory != null) 'subcategory': subcategory,
        'direction': direction,
        'amount': amount,
        'frequency': frequency,
        if (intervalDays != null) 'intervalDays': intervalDays,
        if (nextDueDate != null)
          'nextDueDate': nextDueDate!.toIso8601String().split('T')[0],
        'active': active,
      };

  bool get isIncome => direction == 'INCOME';

  Contract copyWith({
    String? name,
    String? category,
    String? subcategory,
    String? direction,
    double? amount,
    String? frequency,
    int? intervalDays,
    DateTime? nextDueDate,
    bool? active,
  }) {
    return Contract(
      id: id,
      name: name ?? this.name,
      category: category ?? this.category,
      subcategory: subcategory ?? this.subcategory,
      direction: direction ?? this.direction,
      amount: amount ?? this.amount,
      frequency: frequency ?? this.frequency,
      intervalDays: intervalDays ?? this.intervalDays,
      lastBookingDate: lastBookingDate,
      nextDueDate: nextDueDate ?? this.nextDueDate,
      occurrenceCount: occurrenceCount,
      active: active ?? this.active,
      cancelledAt: cancelledAt,
      detectedAutomatically: detectedAutomatically,
      monthlyAmount: monthlyAmount,
    );
  }

  /// Deutsches Etikett des Rhythmus.
  String get frequencyLabel => switch (frequency) {
        'WEEKLY' => 'wöchentlich',
        'BIWEEKLY' => 'zweiwöchentlich',
        'MONTHLY' => 'monatlich',
        'BIMONTHLY' => 'zweimonatlich',
        'QUARTERLY' => 'vierteljährlich',
        'SEMIANNUAL' => 'halbjährlich',
        'YEARLY' => 'jährlich',
        _ => 'unregelmäßig',
      };
}

File: Everything-app-backend/everything-app/src/main/java/com/Finn/everything_app/model/Contract.java
package com.Finn.everything_app.model;

import jakarta.persistence.*;
import lombok.Data;

import java.time.LocalDate;
import java.time.LocalDateTime;

/**
 * Eine wiederkehrende Zahlung - Abo, Miete, Versicherung, aber auch das Gehalt.
 *
 * <p>Loest mittelfristig den Behelf ab, "Vertrag" als Transaktion mit {@code isRecurring = true}
 * zu fuehren. Bis die Oberflaeche umgestellt ist, existieren beide Begriffe nebeneinander.
 *
 * <h2>frequency vs. intervalDays</h2>
 * Die beiden Felder sind redundant, und wer sie verwechselt, baut eine Prognose, die um Wochen
 * danebenliegt:
 * <ul>
 *   <li>{@link #intervalDays} ist der <em>gemessene</em> Median-Abstand zwischen zwei Buchungen
 *       und damit massgeblich fuer jede Datumsrechnung, insbesondere {@link #nextDueDate}.</li>
 *   <li>{@link #frequency} ist das <em>klassifizierte</em> Etikett und massgeblich fuer die
 *       Gruppierung in der Oberflaeche sowie fuer die Monatsnormalisierung der Prognose
 *       (QUARTERLY entspricht {@code amount / 3} pro Monat).</li>
 * </ul>
 *
 * <p>Bewusst ohne Waehrung: {@link BankAccount} fuehrt sie, und eine Waehrung hier wuerde
 * Summenbildungen einen Umrechnungspfad vorgaukeln, den es nirgends gibt.
 */
@Entity
@Table(name = "contracts",
        indexes = {
                @Index(name = "idx_contracts_user", columnList = "user_id"),
                @Index(name = "idx_contracts_user_key", columnList = "user_id, counterparty_key")
        })
@Data
public class Contract {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    /** Anzeigename, vom Nutzer aenderbar. */
    @Column(nullable = false, length = 200)
    private String name;

    /** Normalisierte Gegenpartei (klein, ohne Filialnummern und Datumsfragmente) - der Gruppierungsschluessel der Erkennung. */
    @Column(name = "counterparty_key", nullable = false, length = 200)
    private String counterpartyKey;

    @Column(nullable = false, length = 100)
    private String category;

    @Column(length = 100)
    private String subcategory;

    /**
     * Ausgabe oder Einnahme. Ohne dieses Feld kann die Prognose "Miete 850" nicht von
     * "Gehalt 2400" unterscheiden - und ist damit keine.
     */
    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 20)
    private TransactionType direction = TransactionType.EXPENSE;

    /** Immer ein positiver Betrag; das Vorzeichen traegt {@link #direction}. */
    @Column(nullable = false)
    private Double amount;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 20)
    private ContractFrequency frequency = ContractFrequency.MONTHLY;

    @Column(name = "interval_days")
    private Integer intervalDays;

    @Column(name = "last_booking_date")
    private LocalDate lastBookingDate;

    /** Null bei gekuendigten Vertraegen - Sortierungen muessen das aushalten. */
    @Column(name = "next_due_date")
    private LocalDate nextDueDate;

    /** Anzahl der Buchungen, aus denen der Vertrag erkannt wurde - das Konfidenzsignal der Oberflaeche. */
    @Column(name = "occurrence_count")
    private Integer occurrenceCount = 0;

    @Column
    private Boolean active = true;

    @Column(name = "cancelled_at")
    private LocalDateTime cancelledAt;

    /** Manuell angelegte oder korrigierte Vertraege ({@code false}) darf die Erkennung nicht ueberschreiben. */
    @Column(name = "detected_automatically")
    private Boolean detectedAutomatically = false;

    @Column(name = "created_at")
    private LocalDateTime createdAt;

    @Column(name = "updated_at")
    private LocalDateTime updatedAt;

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "user_id", nullable = false)
    private User user;

    @PrePersist
    protected void onCreate() {
        createdAt = LocalDateTime.now();
        updatedAt = LocalDateTime.now();
    }

    @PreUpdate
    protected void onUpdate() {
        updatedAt = LocalDateTime.now();
    }
}
